src/system/random/basic.lean

```diff
@@ -0,0 +1,479 @@
+    rw one_add, apply zero_lt_succ },
+  { rw [← add_assoc,← nat.add_sub_assoc P',nat.add_sub_cancel_left,add_one],
+    clear P' i',
+    cases y with y Hy,
+    unfold bitvec.to_nat vector.to_list subtype.val bitvec.bits_to_nat,
+    rw [← reverse_reverse y, foldl_reverse,← Hy,← length_reverse],
+    generalize : reverse y = z,
+    clear x' x y' Hy y n,
+    induction z with x xs,
+    { rw [list.length,list.foldr,nat.pow] },
+    { simp [foldr,length,one_add,pow_succ,flip,bitvec.add_lsb],
+      transitivity succ (1 +
+       (foldr (λ (b : bool) (a : ℕ), a + (a + cond b 1 0)) 0 xs +
+          foldr (λ (b : bool) (a : ℕ), a + (a + cond b 1 0)) 0 xs)),
+      { apply succ_le_succ, apply add_le_add_right,
+        cases x, apply nat.zero_le, refl, },
+      { simp!,
+        rw [← nat.add_succ,← nat.add_succ,one_add,← nat.succ_add,mul_comm,← two_mul],
+        apply nat.mul_le_mul_left,
+        simp [flip,bitvec.add_lsb] at z_ih,
+        apply z_ih } }, },
+end
+end coerce
+
+open nat
+
+/-- Use `i` to generate an element of the interval `x .. y` -/
+protected def bitvec.coerce {n : ℕ} (x y : bitvec n) (P : x ≤ y)
+  (i : bitvec n) :
+  (x .. y) :=
+let x' := x.to_nat,
+    y' := y.to_nat,
+    i' := i.to_nat,
+    r := i' % (y' - x' + 1) + x' in
+have Hx : x ≤ bitvec.of_nat n r,
+  begin
+    unfold_locals r,
+    simp [bitvec.le_def,bitvec.to_nat_of_nat],
+    rw [mod_eq_of_lt],
+    { apply nat.le_add_right },
+    unfold_locals x' y' i',
+    apply bitvec.interval_fits_in_word_size,
+    apply P
+  end,
+have Hy : bitvec.of_nat n r ≤ y,
+  begin
+    unfold_locals r,
+    rw [bitvec.le_def,bitvec.to_nat_of_nat,mod_eq_of_lt],
+    transitivity (y' - x') + x',
+    { apply add_le_add_right,
+      apply le_of_lt_succ,
+      rw ← add_one,
+      apply mod_lt,
+      rw add_one, apply zero_lt_succ },
+    { transitivity x' + (y' - x'),
+      apply le_of_eq, ac_refl,
+      rw [← nat.add_sub_assoc P,nat.add_sub_cancel_left], },
+    simp, apply bitvec.interval_fits_in_word_size P,
+  end,
+⟨ bitvec.of_nat _ r , Hx , Hy ⟩
+
+/-- generate an infinite series of bitvectors -/
+protected def bitvec.random_series (n : ℕ) (g : gen) : stream (bitvec n) :=
+stream.corec
+(λ s, ⟨ stream.approx n s, stream.length_approx _ _ ⟩)
+(stream.drop n)
+(@random.random_series bool _ gen _ g)
+
+instance random_bitvec (n : ℕ) : random (bitvec n) :=
+{ to_has_le := by apply_instance,
+  random := λ _ inst, @bitvec.random _ inst n,
+  random_r := λ _ inst x y p, bitvec.coerce _ _ p <$> @bitvec.random _ inst n,
+  random_series := λ _ inst, @bitvec.random_series _ inst n,
+  random_series_r := λ _ inst x y p g, bitvec.coerce _ _ p ∘ @bitvec.random_series _ inst n g }
+
+open nat
+
+lemma div_two_round_up
+  (x : ℕ)
+  (h₀ : 1 < x)
+: (x + 1) / 2 < x :=
+begin
+  rw [div_lt_iff_lt_mul,norm_num.mul_bit0,mul_one,bit0],
+  apply add_lt_add_left, apply h₀,
+  apply of_as_true, trivial
+end
+
+/-- `word_size n` gives us the number of bits required to represent `n` -/
```

Comment:
Huh, I'm surprised we don't have anything about this already. Perhaps put `log` somewhere in the comment, so increase the chance people can find this? And perhaps move to `data.nat.basic`?

Reply:
There's now `data.nat.digits` available.